table/update_schema.go

```diff
@@ -0,0 +1,606 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package table
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/apache/iceberg-go"
+)
+
+type UpdateSchema struct {
+	txn          *Transaction
+	schema       *iceberg.Schema
+	lastColumnID int
+
+	deletes map[int]struct{}
+	updates map[int]*iceberg.NestedField
+	adds    map[int][]*iceberg.NestedField
+	moves   map[int][]moveReq
+
+	allowIncompatibleChanges bool
+	identifierFields         map[string]struct{}
+	caseSensitive            bool
+}
+
+type moveOp string
+
+const (
+	OpFirst  moveOp = "first"
+	OpBefore moveOp = "before"
+	OpAfter  moveOp = "after"
+)
+
+type moveReq struct {
+	fieldID      int
+	otherFieldID int // 0 when opFirst
+	op           moveOp
+}
+
+func NewUpdateSchema(txn *Transaction, s *iceberg.Schema, lastColumnID int) *UpdateSchema {
+	return &UpdateSchema{
+		txn:                      txn,
+		schema:                   s,
+		deletes:                  make(map[int]struct{}),
+		updates:                  make(map[int]*iceberg.NestedField),
+		adds:                     make(map[int][]*iceberg.NestedField),
+		moves:                    make(map[int][]moveReq),
+		lastColumnID:             lastColumnID,
+		allowIncompatibleChanges: false,
+		identifierFields:         make(map[string]struct{}),
+		caseSensitive:            true,
+	}
+}
+
+// AllowIncompatibleChanges permits incompatible schema changes.
+func (us *UpdateSchema) AllowIncompatibleChanges() *UpdateSchema {
+	us.allowIncompatibleChanges = true
+
+	return us
+}
+
+// AddColumn adds a new column to the schema.
+// usage:
+// a is a struct/list/map b is a sub field of a
+//
+// us.AddColumn([]string{"a","b"}, true, iceberg.StringType{}, "doc", "default")
+// us.AddColumn([]string{"a","b"}, true, iceberg.StringType{}, "doc", "default")
+func (us *UpdateSchema) AddColumn(path []string, required bool, dataType iceberg.Type, doc string, initialDefault any) (*UpdateSchema, error) {
```

Comment:
We should either return just `error` or we should return just `*UpdateSchema` and defer all the actual operations to a `Build()` method. 

There's no reason to return *both* because you don't need the returned `*UpdateSchema` to continue, so we are preventing chaining while not providing anything in return. So we should embrace the Builder pattern entirely, or we should just have the single `error` return.